server/embed/serve.go

```diff
@@ -109,113 +111,78 @@ func (sctx *serveCtx) serve(
-		sctx.lg.Info(
-			"serving client traffic insecurely; this is strongly discouraged!",
-			zap.String("address", sctx.l.Addr().String()),
-		)
+	gs := v3rpc.Server(s, nil, nil, gopts...)
+	defer func() {
+		sctx.lg.Warn("stopping grpc server due to error", zap.Error(err))
+		gs.Stop()
+		sctx.lg.Warn("stopped grpc server due to error", zap.Error(err))
+	}()
+	v3electionpb.RegisterElectionServer(gs, servElection)
+	v3lockpb.RegisterLockServer(gs, servLock)
+	if sctx.serviceRegister != nil {
+		sctx.serviceRegister(gs)
 	}
-
+	var gwmux *gw.ServeMux
+	var tlscfg *tls.Config
 	if sctx.secure {
-		tlscfg, tlsErr := tlsinfo.ServerConfig()
-		if tlsErr != nil {
-			return tlsErr
+		tlscfg, err = tlsinfo.ServerConfig()
+		if err != nil {
+			return err
 		}
-		gs = v3rpc.Server(s, tlscfg, nil, gopts...)
-		v3electionpb.RegisterElectionServer(gs, servElection)
-		v3lockpb.RegisterLockServer(gs, servLock)
-		if sctx.serviceRegister != nil {
-			sctx.serviceRegister(gs)
+	}
+	if s.Cfg.EnableGRPCGateway {
+		var creds credentials.TransportCredentials
+		if sctx.insecure {
+			creds = insecure.NewCredentials()
 		}
-		handler = grpcHandlerFunc(gs, handler)
-
-		var gwmux *gw.ServeMux
-		if s.Cfg.EnableGRPCGateway {
-			dtls := tlscfg.Clone()
-			// trust local server
-			dtls.InsecureSkipVerify = true
-			bundle := credentials.NewBundle(credentials.Config{TLSConfig: dtls})
-			opts := []grpc.DialOption{grpc.WithTransportCredentials(bundle.TransportCredentials())}
-			gwmux, err = sctx.registerGateway(opts)
-			if err != nil {
-				return err
+		if sctx.secure {
+			if s.Cfg.EnableGRPCGateway {
+				dtls := tlscfg.Clone()
+				// trust local server
+				dtls.InsecureSkipVerify = true
+				bundle := clientcreds.NewBundle(clientcreds.Config{TLSConfig: dtls})
+				creds = bundle.TransportCredentials()
 			}
 		}
-
-		var tlsl net.Listener
-		tlsl, err = transport.NewTLSListener(m.Match(cmux.Any()), tlsinfo)
+		gwmux, err = sctx.registerGateway([]grpc.DialOption{grpc.WithTransportCredentials(creds)})
 		if err != nil {
+			sctx.lg.Error("registerGateway failed", zap.Error(err))
 			return err
 		}
-		// TODO: add debug flag; enable logging when debug flag is set
-		httpmux := sctx.createMux(gwmux, handler)
+	}
+	// TODO: add debug flag; enable logging when debug flag is set
+	srv := &http.Server{
+		Handler:  createAccessController(sctx.lg, s, sctx.createMux(gwmux, handler)),
+		ErrorLog: logger, // do not log user error
+	}
+	if err := configureHttpServer(srv, s.Cfg); err != nil {
+		sctx.lg.Error("Configure https server failed", zap.Error(err))
+		return err
+	}
 
-		srv := &http.Server{
-			Handler:   createAccessController(sctx.lg, s, httpmux),
-			TLSConfig: tlscfg,
-			ErrorLog:  logger, // do not log user error
-		}
-		if err := configureHttpServer(srv, s.Cfg); err != nil {
-			sctx.lg.Error("Configure https server failed", zap.Error(err))
+	listener := sctx.l
+	if sctx.secure {
+		listener, err = transport.NewTLSListener(listener, tlsinfo)
```

Comment:
Are we covering here /healthz and /metrics endpoints with mTLS that used to be unprotected previously ?

https://github.com/etcd-io/etcd/issues/11993
https://github.com/etcd-io/etcd/pull/12370#issuecomment-714676715

Reply:
Don't think so. There is no healthz or metrics handler that I'm touching there. The only existing handler in serve was under TLS and will be under TLS.